Add support for Gitlab CI

wmanley opened the pull request:
Created based on the documentation here: https://docs.gitlab.com/ee/ci/variables/ .  Not tested with real gitlab CI

TODO:

- [x] Test with gitlab CI.
- [x] Find out what test-reporting format it uses and provide that. JUnit XML: https://docs.gitlab.com/ee/ci/yaml/artifacts_reports.html#artifactsreportsjunit
- [x] Check and document password/token handling: Specify STBT_AUTH_TOKEN in the GitLab UI, mark it as "masked", then reassign them in your `.gitlab-ci.yml`:

    ```
    variables:
      STBT_AUTH_TOKEN: $STBT_AUTH_TOKEN
    ```

    See https://docs.gitlab.com/ee/ci/variables/

- [x] See what happens when you stop a job: SIGTERM on Unix, SIGKILL on Windows. https://docs.gitlab.com/runner/executors/shell.html#gitlab-131-and-later
- [ ] Document gitlab CI setup

kalidasya commented on stbt_rig.py:
```diff
@@ -100,6 +100,27 @@ def main(argv):
+              If so, we enable the following behaviours:
+
+              * Read the access token from $CI_BUILD_TOKEN environment variable.
+              * Record various Jenkins parameters as "tags" in the Stb-tester
```
typo, Jenkins -> GitlabCI

drothlis commented:
> TODO: HOW DO RESULTS WORK IN GITLAB CI?

JUnit XML: https://docs.gitlab.com/ee/ci/yaml/artifacts_reports.html#artifactsreportsjunit

drothlis commented:
> TODO: HOW DOES STOPPING JOBS WORK IN GITLAB CI?

SIGTERM on Unix, SIGKILL on Windows:
https://docs.gitlab.com/runner/executors/shell.html#gitlab-131-and-later

drothlis commented on stbt_rig.py:
```diff
@@ -329,6 +356,25 @@ def cmd_run(args, node):
+        # 2017-03-22).  We support both sets of names, but always record the
+        # new name.
+        # See https://docs.gitlab.com/ee/ci/variables/
+        #                    8.x name      9.0+ name
+        for k, old_key in [('CI_JOB_ID', 'CI_BUILD_ID'),
```
I think the comment is the wrong way round. The current [gitlab docs](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html#predefined-variables-reference) list CI_JOB_* but not CI_BUILD_*.

drothlis replied:
Ah yes, the code is correct, just the comment is wrong.

drothlis commented on stbt_rig.py:
```diff
@@ -400,6 +446,14 @@ def iter_portal_auth_tokens(portal_url, portal_auth_file, mode):
         return
 
+    if mode == "gitlab":
+        token = os.environ.get("CI_BUILD_TOKEN")
```
GitLab's CI_BUILD_TOKEN (now called CI_JOB_TOKEN) is to access the GitLab API during the job. It isn't to access the Stb-tester Portal.

drothlis committed:
Add support for GitLab CI

Based on the documentation here:

- https://docs.gitlab.com/ee/ci/variables/
- https://docs.gitlab.com/ee/ci/yaml/artifacts_reports.html#artifactsreportsjunit
- https://docs.gitlab.com/runner/executors/shell.html#gitlab-131-and-later

drothlis commented:
Rebased onto "main" to fix conflicts. I dropped the compatibility code for GitLab 8 (2017) and addressed the TODO items (just from reading the docs, I don't have a GitLab instance to test against).

drothlis commented:
Tested by @kalidasya — thanks!
